### PR TITLE
Epoch2

### PR DESCRIPTION
**Pull request by nickdesaulniers**
this PR implements "epoch2" in which the results from stage2's build from last night becomes the new "stage0" today.

Retain "epoch1" for when we want to bootstrap from a non-x86_64 host.

### CODE REVIEW

**Comment by nathanchance** [comment]
This might sound like a stupid or strongly worded question (I don't intend for it to be) but I want to make sure I understand what is happening here so that we aren't duplicating anything or doing certain things unnecessarily: what's the point of rebuilding stage 2 with stage 2 when we can just build stage 3 through 5 and use stage 5 as stage 2? Is this just to make sure that stage 2 can build all the future stages or is this laying the foundation to turn the "new" stage 2 into stage 3?

If it is the former, please call it out with a comment somewhere, because on the surface, it seems somewhat pointless to build stage 2 twice.

If it is the latter, I think we would just be better off calling this stage 3 but labelling it as a WIP.

If it is something else, then ignore my comments :)

**Comment by nickdesaulniers** [comment]
Good questions, they are a forcing function for me to think harder about the problem, get the ideas out of my head, and clarify my expressions.

> what's the point of rebuilding stage 2 with stage 2 when we can just build stage 3 through 5 and use stage 5 as stage 2?

We totally _could_, that is an option.

I prefer to get this thing bootstrapping itself ASAP though, and that option would delay that until as late as possible.  That way the bootstrapping is in place in CI for when we implement stage3, then stage 4, then stage 5.

So once stage3 is implemented, I want the "second" stage3 build built via stage3 from the prior epoch. (And so on for stage4, then stage 5, piecemeal).

> Is this just to make sure that stage 2 can build all the future stages or is this laying the foundation to turn the "new" stage 2 into stage 3?

Mostly the latter.

> If it is the latter, I think we would just be better off calling this stage 3 but labelling it as a WIP.

To me, the output at this point is not different other than what toolchain was used to build it; previously stage 2 was built with stage 1. Now stage 2 (epoch2) is built with stage 2 (from epoch1; called `prev_epoch` in the Dockerfile).  Stage 3 will be built entirely from sources; this is not.

Whether we call that something else ("stage 3 WIP", "stage 2 (epoch 2)", "stage 2b"), sure, we can do that. But I don't want to classify this as "stage 3" until it's fully built from sources (minus building cmake+ninja+make from sources).  I wouldn't call it a distinct stage per se when swapping to a new epoch.

(Maybe "epoch N-1" is more descriptive?)

**Review by nathanchance — COMMENTED** [review]
Okay, I think that all makes sense to me. So basically, we want stage 1 and 2 to be irrelevant once stage 3 is finalized, as that is basically what we want to ship, so we are going to build the second stage 2 with the first stage 2 to work on getting the second stage 2 to the first stage 3 then do the same thing for stage 3 and so on and so forth. If I am missing something, please let me know!

I wonder if it would be better to split up the workflow files so that we are not tied to rebuilding epoch one when we want to rebuild epoch 1, as it seems like epoch 2 is going to have more to it than just the stage 2 rebuild later?

**Comment by nickdesaulniers** [comment]
> So basically, we want stage 1 and 2 to be irrelevant once stage 3 is finalized

Yeah, that paragraph sounds right to me.

> it seems like epoch 2 is going to have more to it than just the stage 2 rebuild later?

Right, perhaps a workflow per epoch is how to split things up?  I'm not sure yet.

> so that we are not tied to rebuilding epoch one when we want to rebuild epoch 1

I don't think we'll want to rebuild epoch 1 for x86_64 hosts ever again.  For now, even for upgrades of the base llvm version, we should be using the stage2 container for the initial toolchain.

The split between epoch files is so that we can bootstrap via epoch 1 on other host architectures, say aarch64.  I figure we can make slight modifications to the Dockerfile.epoch1 to make it more arch agnostic, then use that for bootstrapping an aarch64 host build.

I can probably intersperse the documentation about epochs so that they read inline with each stage. That would make it clearer when a change in epochs is expected or makes sense.

> I wonder if it would be better to split up the workflow files

Probably, though I'm having trouble imagining what you are.  What are your thoughts on how we can split up the workflow files?

**Comment by nickdesaulniers** [comment]
> I don't think we'll want to rebuild epoch 1 for x86_64 hosts ever again.

I guess the tradeoff is do we want to continuously test that we can rebootstrap from alpine through sucessive releases of llvm?  Perhaps, but that's not _the_ main goal of this work (which is to produce releases), more of a stretch goal or icing on the cake.

**Comment by nathanchance** [comment]
> Right, perhaps a workflow per epoch is how to split things up? I'm not sure yet.

Yes, this is what I was thinking because...

> I don't think we'll want to rebuild epoch 1 for x86_64 hosts ever again. For now, even for upgrades of the base llvm version, we should be using the stage2 container for the initial toolchain.

with the way the workflow is wired up, we are going to build

* stage 1
* stage 2 with stage 1
* stage 2 with stage 2

but we don't want the first two steps every time, correct? We only want the first two steps to happen if we don't have a stage 2, so we should probably do `mv llvm-project.yml llvm-project-epoch-one.yml` and add your proposed changes to `llvm-project-epoch-two.yml` or whatever.

Alternatively, we could keep everything in one workflow by [adding inputs to the `workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_dispatchinputs) to control exactly what we build. This would work for now but that becomes fragile if we ever want to start running this in a more automated fashion.

The inputs look like:

image

https://github.com/nathanchance/env/blob/2d2324b2472abffc968214108740284dbe213870/.github/workflows/docker-dev.yml#L10-L15

**Comment by nickdesaulniers** [comment]
> but we don't want the first two steps every time, correct?

Right.

> so we should probably do mv llvm-project.yml llvm-project-epoch-one.yml and add your proposed changes to llvm-project-epoch-two.yml or whatever.

I see.  I guess you already had me convinced when I said

> Right, perhaps a workflow per epoch is how to split things up?

I just need to actually implement that! Will do so now.

File: .github/workflows/llvm-project-epoch-one.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+name: Build kernel.org toolchain
```
Comment (nathanchance) [inline .github/workflows/llvm-project-epoch-one.yml:1]: Make this unique between the two workflows?

**Review by nathanchance — APPROVED** [review]
Yes, I think this is what we want! The final diff looks fine to me.

File: .github/workflows/llvm-project-epoch-one.yml (hunk 1 above)
Comment (nickdesaulniers) [reply .github/workflows/llvm-project-epoch-one.yml]: done PTAL

**Review by nathanchance — APPROVED** [review]
Let's see what explodes :^)

**Pull request merged**